# 1. Business Understanding

#### In This project, I will Answer Three Quations, there Are:
#####     Q1: How to find the hight of station with standerd error ..?
#####     Q2: Can you see if temperature and humidity have changed or not along 11 years ..?
#####     Q3: Is there correlation between temperature and humidity ..? 

# 2. Data Understanding

The data used is weather data for the city of Mecca in the Kingdom of Saudi Arabia, covering the period from 2008 to 2017.

# 3. Prepare Data

## 3.1 Import Libraries:

In this section, I will import all library it will be  useful.

In [3]:
import pandas as pd    
import numpy as np
import plotly.express as px
import seaborn as sns

## 3.2 Create DataFrame:

#### In this section, I will create a dataframe of my study.

In [4]:
# df = pd.read_csv('C:/users/Ibrahim Al-Khurayyif/Project1/Data_M2008-2017.csv')
df = pd.read_csv('C:/Users/Ibrahim Al Khurayyif/Udacity - Data Scientist Nanodegree/Project_1/Project1/Data_M2008-2017.csv')

In [5]:
df

STATION_NUMBER STATION_NAME  YEAR  MONTH  DAY  AIR_TEMPERATURE_MEAN_DB  \
0              41030       Makkah  2008      1    1                     26.7   
1              41030       Makkah  2008      1    2                     26.4   
2              41030       Makkah  2008      1    3                     26.2   
3              41030       Makkah  2008      1    4                     25.2   
4              41030       Makkah  2008      1    5                     23.8   
...              ...          ...   ...    ...  ...                      ...   
3578           41030       Makkah  2017     12   26                     27.2   
3579           41030       Makkah  2017     12   27                     26.6   
3580           41030       Makkah  2017     12   28                     27.4   
3581           41030       Makkah  2017     12   29                     27.8   
3582           41030       Makkah  2017     12   30                     27.6   

      AIR_TEMPERATURE_MAX_DB  AIR_TEMPERATURE_MIN_DB  \
0                       33.8                    21.0   
1                       33.8                    21.0   
2                       33.0                    22.0   
3                       30.2                    22.4   
4                       30.4                    21.0   
...                      ...                     ...   
3578                    37.0                    21.2   
3579                    35.0                    20.6   
3580                    35.6                    22.6   
3581                    35.6                    23.0   
3582                    34.0                    23.4   

     WIND_PREVAILING_DIRECTION  WIND_MEAN_SPEED  ...  RELATIVE_HUMIDITY_MEAN  \
0                            S              2.0  ...                      65   
1                          WSW              1.0  ...                      65   
2                          WSW              1.0  ...                      72   
3                            N              1.0  ...                      65   
4                           NW              1.0  ...                      57   
...                        ...              ...  ...                     ...   
3578                      CALM              0.0  ...                      53   
3579                         S              2.0  ...                      51   
3580                       SSW              1.0  ...                      63   
3581                       WSW              2.0  ...                      64   
3582                       SSW              2.0  ...                      67   

      SKY_COV_OKTES_MEAN  RELATIVE_HUMIDITY_MAX  RELATIVE_HUMIDITY_MIN  \
0                    0.6                     90                     39   
1                    0.3                     87                     38   
2                    1.9                     92                     42   
3                    3.5                     89                     48   
4                    3.4                     66                     42   
...                  ...                    ...                    ...   
3578                 1.3                     75                     24   
3579                 0.1                     81                     38   
3580                 0.2                     84                     33   
3581                 0.5                     85                     32   
3582                 0.4                     86                     39   

      PRESSURE_MIN_SEA_LEVEL  PRESSURE_MIN_STATION_LEVEL  \
0                     1009.0                       982.0   
1                     1010.3                       983.0   
2                     1010.2                       983.3   
3                     1012.3                       985.3   
4                     1013.3                       986.0   
...                      ...                         ...   
3578                  1010.9                       983.4   
3579                  1010.5              

## 3.3 Prepare DataFrame:

#### In this section, I will do two things:

#####    (A) Change the column name formatting style to "Capitalization".

#####    (B) Drop columns that will don't use them.

#####    (C) Create columns that will be useful for study.

#####    (D) Reorder columns.

### 3.3.A Capitalization:

#### I will do that because the old column name formatting style is not comfortable.

In [6]:
df.columns = ['StationNumber', 'StationName', 'Year', 'Month', 'Day',
       'AirTemperatureMeanDB', 'AirTemperatureMaxDB',
       'AirTemperatureMinDB', 'WindPrevailingDirection',
       'WindMeanSpeed', 'WindMaxDirection', 'WindMaxSpeed',
       'RelativeHumidityMean', 'SkyCovOktesMean', 'RelativeHumidityMax',
       'RelativeHumidityMin', 'PressureMinSeaLevel',
       'PressureMinStationLevel', 'PressureMaxStationLevel',
       'PressureMaxSeaLevel', 'PressureMeanStationLevel',
       'PressureMeanSeaLevel']

In [7]:
df.columns

Index(['StationNumber', 'StationName', 'Year', 'Month', 'Day',
       'AirTemperatureMeanDB', 'AirTemperatureMaxDB', 'AirTemperatureMinDB',
       'WindPrevailingDirection', 'WindMeanSpeed', 'WindMaxDirection',
       'WindMaxSpeed', 'RelativeHumidityMean', 'SkyCovOktesMean',
       'RelativeHumidityMax', 'RelativeHumidityMin', 'PressureMinSeaLevel',
       'PressureMinStationLevel', 'PressureMaxStationLevel',
       'PressureMaxSeaLevel', 'PressureMeanStationLevel',
       'PressureMeanSeaLevel'],
      dtype='object')

### 3.3.B Drop Columns:

#### In this section, I will drop some columns like: StationNumber, StationName or any columns is will not use.

In [8]:
df = df.drop(columns=['StationNumber', 'StationName'])

### 3.3.C Create A New Column:

#### In this section, I will use 'to_datetime' tool from pandas to create a new column.

In [9]:
df['Date'] = pd.to_datetime(df["Year"].astype(str) + "-" + df["Month"].astype(str) + "-" + df["Day"].astype(str))

In [10]:
df

Year  Month  Day  AirTemperatureMeanDB  AirTemperatureMaxDB  \
0     2008      1    1                  26.7                 33.8   
1     2008      1    2                  26.4                 33.8   
2     2008      1    3                  26.2                 33.0   
3     2008      1    4                  25.2                 30.2   
4     2008      1    5                  23.8                 30.4   
...    ...    ...  ...                   ...                  ...   
3578  2017     12   26                  27.2                 37.0   
3579  2017     12   27                  26.6                 35.0   
3580  2017     12   28                  27.4                 35.6   
3581  2017     12   29                  27.8                 35.6   
3582  2017     12   30                  27.6                 34.0   

      AirTemperatureMinDB WindPrevailingDirection  WindMeanSpeed  \
0                    21.0                       S            2.0   
1                    21.0                     WSW            1.0   
2                    22.0                     WSW            1.0   
3                    22.4                       N            1.0   
4                    21.0                      NW            1.0   
...                   ...                     ...            ...   
3578                 21.2                    CALM            0.0   
3579                 20.6                       S            2.0   
3580                 22.6                     SSW            1.0   
3581                 23.0                     WSW            2.0   
3582                 23.4                     SSW            2.0   

      WindMaxDirection  WindMaxSpeed  ...  SkyCovOktesMean  \
0                   19            10  ...              0.6   
1                   18             7  ...              0.3   
2                   23             9  ...              1.9   
3                   31             6  ...              3.5   
4                   27             5  ...              3.4   
...                ...           ...  ...              ...   
3578                24             4  ...              1.3   
3579                18             8  ...              0.1   
3580                24             6  ...              0.2   
3581                18            10  ...              0.5   
3582                21             8  ...              0.4   

      RelativeHumidityMax  RelativeHumidityMin  PressureMinSeaLevel  \
0                      90                   39               1009.0   
1                      87                   38               1010.3   
2                      92                   42               1010.2   
3                      89                   48               1012.3   
4                      66                   42               1013.3   
...                   ...                  ...                  ...   
3578                   75                   24               1010.9   
3579                   81                   38               1010.5   
3580                   84                   33               1010.4   
3581                   85                   32               1011.4   
3582                   86                   39               1011.6   

      PressureMinStationLevel  PressureMaxStationLevel  PressureMaxSeaLevel  \
0                       982.0                    986.4               1013.4   
1                       983.0                    986.3               1013.1   
2                       983.3                    986.8               1013.9   
3                       985.3                    988.7               1015.6   
4                       986.0                    988.9               1016.0   
...                       ...                      ...                  ...   
3578                    983.4                    988.5               1015.8   
3579                    983.5                    987.6               1014.5   
3580                    983.7                    98

### 3.3.D Reorder The Columns:

#### In this section, I will reorder columns to facilitate reading.

In [11]:
df = df[['Date', 'AirTemperatureMeanDB', 'AirTemperatureMaxDB',
       'AirTemperatureMinDB', 'WindPrevailingDirection', 'WindMeanSpeed',
       'WindMaxDirection', 'WindMaxSpeed', 'RelativeHumidityMean',
       'SkyCovOktesMean', 'RelativeHumidityMax', 'RelativeHumidityMin',
       'PressureMinSeaLevel', 'PressureMinStationLevel',
       'PressureMaxStationLevel', 'PressureMaxSeaLevel',
       'PressureMeanStationLevel', 'PressureMeanSeaLevel', 'Year', 'Month', 'Day']]

In [12]:
df

Date  AirTemperatureMeanDB  AirTemperatureMaxDB  \
0    2008-01-01                  26.7                 33.8   
1    2008-01-02                  26.4                 33.8   
2    2008-01-03                  26.2                 33.0   
3    2008-01-04                  25.2                 30.2   
4    2008-01-05                  23.8                 30.4   
...         ...                   ...                  ...   
3578 2017-12-26                  27.2                 37.0   
3579 2017-12-27                  26.6                 35.0   
3580 2017-12-28                  27.4                 35.6   
3581 2017-12-29                  27.8                 35.6   
3582 2017-12-30                  27.6                 34.0   

      AirTemperatureMinDB WindPrevailingDirection  WindMeanSpeed  \
0                    21.0                       S            2.0   
1                    21.0                     WSW            1.0   
2                    22.0                     WSW            1.0   
3                    22.4                       N            1.0   
4                    21.0                      NW            1.0   
...                   ...                     ...            ...   
3578                 21.2                    CALM            0.0   
3579                 20.6                       S            2.0   
3580                 22.6                     SSW            1.0   
3581                 23.0                     WSW            2.0   
3582                 23.4                     SSW            2.0   

      WindMaxDirection  WindMaxSpeed  RelativeHumidityMean  SkyCovOktesMean  \
0                   19            10                    65              0.6   
1                   18             7                    65              0.3   
2                   23             9                    72              1.9   
3                   31             6                    65              3.5   
4                   27             5                    57              3.4   
...                ...           ...                   ...              ...   
3578                24             4                    53              1.3   
3579                18             8                    51              0.1   
3580                24             6                    63              0.2   
3581                18            10                    64              0.5   
3582                21             8                    67              0.4   

      ...  RelativeHumidityMin  PressureMinSeaLevel  PressureMinStationLevel  \
0     ...                   39               1009.0                    982.0   
1     ...                   38               1010.3                    983.0   
2     ...                   42               1010.2                    983.3   
3     ...                   48               1012.3                    985.3   
4     ...                   42               1013.3                    986.0   
...   ...                  ...                  ...                      ...   
3578  ...                   24               1010.9                    983.4   
3579  ...                   38               1010.5                    983.5   
3580  ...                   33               1010.4                    983.7   
3581  ...                   32               1011.4                    984.7   
3582  ...                   39               1011.6                    984.7   

      PressureMaxStationLevel  PressureMaxSeaLevel  PressureMeanStationLevel  \
0                       986.4               1013.4                     984.2   
1                       986.3               1013.1                     984.7   
2                       986.8               1013.9                     985.0   
3                       988.7               1015.6                     986.7   
4                       988.9               1016.0                     987.6   
...                       ...                  ...         

#### Now I Finished The Third Process of CRISP-DM (Prepare Data).

# 4. Model Data

#### Q1: How to find the hight of station with standerd error ..?

##### Firstly, I will find mean of AirTemperatureMean by Month:

In [104]:
Temperature = df.groupby(['Year', 'Month']).agg({'AirTemperatureMeanDB':'mean'}).reset_index()

In [105]:
Temperature

Year  Month  AirTemperatureMeanDB
0    2008      1             23.560000
1    2008      2             25.555172
2    2008      3             29.323333
3    2008      4             32.443333
4    2008      5             34.273333
..    ...    ...                   ...
115  2017      8             36.376667
116  2017      9             36.070000
117  2017     10             33.333333
118  2017     11             29.420000
119  2017     12             27.456667

[120 rows x 3 columns]

In [106]:
Temperature = Temperature.pivot(index='Year', columns='Month', values='AirTemperatureMeanDB')
Temperature.columns = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
Temperature = Temperature.reset_index()

In [107]:
Temperature

Year        Jan        Feb        Mar        Apr        May        Jun  \
0  2008  23.560000  25.555172  29.323333  32.443333  34.273333  36.203333   
1  2009  25.813333  26.782143  27.296667  31.923333  33.880000  36.983333   
2  2010  25.326667  25.846429  28.540000  31.993333  34.363333  36.783333   
3  2011  24.010000  25.846429  26.880000  30.646667  34.356667  35.936667   
4  2012  23.553333  26.572414  27.296667  31.283333  35.510000  37.386667   
5  2013  25.276667  27.314286  29.583333  30.826667  34.986667  36.983333   
6  2014  25.650000  26.471429  28.780000  32.706667  34.053333  36.946667   
7  2015  24.416667  26.560714  29.403333  30.836667  35.196667  35.966667   
8  2016  24.396667  26.765517  31.540000  31.676667  35.530000  38.876667   
9  2017  27.516667  26.178571  29.386667  33.320000  35.313333  37.013333   

         Jul        Aug        Sep        Oct        Nov        Dec  
0  36.783333  36.166667  35.600000  32.430000  28.816667  26.706667  
1  36.316667  35.736667  34.920000  32.153333  28.380000  25.493333  
2  36.180000  36.666667  35.263333  34.290000  29.996667  26.833333  
3  36.853333  35.843333  34.703333  32.826667  27.500000  26.413333  
4  36.866667  36.163333  35.323333  32.576667  29.310000  26.460000  
5  36.070000  34.963333  35.466667  33.046667  29.636667  25.913333  
6  36.890000  36.523333  35.096667  32.596667  28.440000  27.310000  
7  36.986667  37.386667  36.150000  33.793333  29.580000  25.246667  
8  37.970000  36.753333  35.846667  33.236667  29.810000  28.240000  
9  37.476667  36.376667  36.070000  33.333333  29.420000  27.456667

###### Secondly, I will find mean of RelativeHumidityMean by Month:

In [108]:
Humidity = df.groupby(['Year', 'Month']).agg({'RelativeHumidityMean': 'mean'}).reset_index()

In [109]:
Humidity = Humidity.pivot(index='Year', columns='Month', values='RelativeHumidityMean')
Humidity.columns = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
Humidity = Humidity.reset_index()

In [110]:
Humidity

Year        Jan        Feb        Mar        Apr        May        Jun  \
0  2008  58.133333  50.758621  42.666667  39.766667  31.333333  28.166667   
1  2009  53.766667  55.964286  50.100000  38.066667  35.433333  30.766667   
2  2010  57.933333  62.214286  51.433333  46.400000  38.233333  34.233333   
3  2011  60.066667  53.714286  46.833333  39.833333  33.533333  33.500000   
4  2012  55.133333  52.862069  44.566667  40.233333  30.566667  24.900000   
5  2013  56.066667  51.642857  47.933333  39.800000  36.133333  24.566667   
6  2014  59.233333  55.428571  47.533333  42.266667  39.933333  35.433333   
7  2015  59.666667  58.750000  52.100000  41.733333  41.333333  35.166667   
8  2016  63.400000  60.931034  52.800000  52.600000  47.533333  35.300000   
9  2017  56.000000  49.142857  48.300000  41.133333  36.433333  35.100000   

         Jul        Aug        Sep        Oct        Nov        Dec  
0  25.433333  31.766667  45.566667  44.266667  57.900000  50.666667  
1  29.533333  41.000000  42.600000  50.166667  61.733333  66.200000  
2  44.366667  43.266667  44.633333  38.566667  48.800000  53.633333  
3  33.066667  36.700000  36.400000  49.800000  49.300000  52.933333  
4  31.800000  33.833333  33.933333  49.133333  50.200000  54.366667  
5  38.366667  43.033333  44.633333  50.333333  56.466667  58.500000  
6  32.733333  39.200000  44.533333  53.666667  59.566667  66.000000  
7  29.466667  44.000000  48.033333  50.666667  58.200000  58.100000  
8  39.300000  48.100000  51.966667  56.166667  60.200000  56.100000  
9  30.866667  45.233333  43.766667  47.966667  59.300000  52.700000

###### Thirdly, I will calculate the annual average of temperature and humidity. Then, I will find the difference between each monthly average for both variables and the overall average:

In [150]:
Temp_mean = df.groupby('Year').agg({'AirTemperatureMeanDB': 'mean'}).reset_index()
Temp_mean.columns = ['Year', 'AnnualTemperatureAverage']
Temp_mean

Year  AnnualTemperatureAverage
0  2008                 31.505014
1  2009                 31.331844
2  2010                 31.873743
3  2011                 31.013408
4  2012                 31.538997
5  2013                 31.696648
6  2014                 31.818436
7  2015                 31.822905
8  2016                 32.569638
9  2017                 32.439944

In [152]:
Humi_mean = df.groupby('Year').agg({'RelativeHumidityMean': 'mean'}).reset_index()
Humi_mean.columns = ['Year', 'AnnualRelativeHumidityAverage']
Humi_mean

Year  AnnualRelativeHumidityAverage
0  2008                      42.178273
1  2009                      46.223464
2  2010                      46.891061
3  2011                      43.751397
4  2012                      41.763231
5  2013                      45.589385
6  2014                      47.918994
7  2015                      48.041899
8  2016                      52.008357
9  2017                      45.474860

In [182]:
Temp_delta = Temperature.copy()
Humi_delta = Humidity.copy()

In [183]:
Temp_delta = Temp_delta.iloc[:, 1:] - Temp_mean['AnnualTemperatureAverage'].values[0]
Temp_delta['Year'] = Temp_mean['Year']
Temp_delta = Temp_delta[['Year', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']]
Temp_delta

Year       Jan       Feb       Mar       Apr       May       Jun       Jul  \
0  2008 -7.945014 -5.949842 -2.181681  0.938319  2.768319  4.698319  5.278319   
1  2009 -5.691681 -4.722871 -4.208347  0.418319  2.374986  5.478319  4.811653   
2  2010 -6.178347 -5.658585 -2.965014  0.488319  2.858319  5.278319  4.674986   
3  2011 -7.495014 -5.658585 -4.625014 -0.858347  2.851653  4.431653  5.348319   
4  2012 -7.951681 -4.932600 -4.208347 -0.221681  4.004986  5.881653  5.361653   
5  2013 -6.228347 -4.190728 -1.921681 -0.678347  3.481653  5.478319  4.564986   
6  2014 -5.855014 -5.033585 -2.725014  1.201653  2.548319  5.441653  5.384986   
7  2015 -7.088347 -4.944300 -2.101681 -0.668347  3.691653  4.461653  5.481653   
8  2016 -7.108347 -4.739497  0.034986  0.171653  4.024986  7.371653  6.464986   
9  2017 -3.988347 -5.326442 -2.118347  1.814986  3.808319  5.508319  5.971653   

        Aug       Sep       Oct       Nov       Dec  
0  4.661653  4.094986  0.924986 -2.688347 -4.798347  
1  4.231653  3.414986  0.648319 -3.125014 -6.011681  
2  5.161653  3.758319  2.784986 -1.508347 -4.671681  
3  4.338319  3.198319  1.321653 -4.005014 -5.091681  
4  4.658319  3.818319  1.071653 -2.195014 -5.045014  
5  3.458319  3.961653  1.541653 -1.868347 -5.591681  
6  5.018319  3.591653  1.091653 -3.065014 -4.195014  
7  5.881653  4.644986  2.288319 -1.925014 -6.258347  
8  5.248319  4.341653  1.731653 -1.695014 -3.265014  
9  4.871653  4.564986  1.828319 -2.085014 -4.048347

In [184]:
Humi_delta = Humi_delta.iloc[:, 1:] - Humi_mean['AnnualRelativeHumidityAverage'].values[0]
Humi_delta['Year'] = Humi_mean['Year']
Humi_delta = Humi_delta[['Year', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']]
Humi_delta

Year        Jan        Feb        Mar        Apr        May        Jun  \
0  2008  15.955060   8.580348   0.488394  -2.411606 -10.844940 -14.011606   
1  2009  11.588394  13.786013   7.921727  -4.111606  -6.744940 -11.411606   
2  2010  15.755060  20.036013   9.255060   4.221727  -3.944940  -7.944940   
3  2011  17.888394  11.536013   4.655060  -2.344940  -8.644940  -8.678273   
4  2012  12.955060  10.683796   2.388394  -1.944940 -11.611606 -17.278273   
5  2013  13.888394   9.464584   5.755060  -2.378273  -6.044940 -17.611606   
6  2014  17.055060  13.250298   5.355060   0.088394  -2.244940  -6.744940   
7  2015  17.488394  16.571727   9.921727  -0.444940  -0.844940  -7.011606   
8  2016  21.221727  18.752762  10.621727  10.421727   5.355060  -6.878273   
9  2017  13.821727   6.964584   6.121727  -1.044940  -5.744940  -7.078273   

         Jul        Aug       Sep        Oct        Nov        Dec  
0 -16.744940 -10.411606  3.388394   2.088394  15.721727   8.488394  
1 -12.644940  -1.178273  0.421727   7.988394  19.555060  24.021727  
2   2.188394   1.088394  2.455060  -3.611606   6.621727  11.455060  
3  -9.111606  -5.478273 -5.778273   7.621727   7.121727  10.755060  
4 -10.378273  -8.344940 -8.244940   6.955060   8.021727  12.188394  
5  -3.811606   0.855060  2.455060   8.155060  14.288394  16.321727  
6  -9.444940  -2.978273  2.355060  11.488394  17.388394  23.821727  
7 -12.711606   1.821727  5.855060   8.488394  16.021727  15.921727  
8  -2.878273   5.921727  9.788394  13.988394  18.021727  13.921727  
9 -11.311606   3.055060  1.588394   5.788394  17.121727  10.521727

##### Finally, I will calculate the hight of Station above sea level by this Physical equation:

I created function called hight, It describe How to find the hight of station, I relied on these columns: PressureMeanStationLevel, PressureMeanSeaLevel and AirTemperatureMeanDB
To find it, and I will calculate the mean of hight.

In [185]:
def Hight(df, PreStat_col, PreSea_col, Temp_col):
    
    '''
    ++++++++++++++++++++++++++++++++++
    |||     Barometric Formula     |||
    ++++++++++++++++++++++++++++++++++
    
        This Function besed on Barometric Formula, It describes the Relationshep between Hight and Pressure:
            dP/dh = -ρ_air * g 
        Where:
            ρ_air = (P * M_ / (R * T)    Density of air
            g = 9.80665 m / sec ** 2     Gravitational acceleration
            R =  8.314 J/K * mol         Universal gas constant
            M = 0.02896 K/mol            Molar mass of air
            T: Tempreture with Kelvin unit   [T_Kelin = 273.15 + T_Celsius]
            

        So, The General Equaton is:
                P(h_station) =  P(h=0) * exp(- M * g  h_station / (R * T))
            That means:
                h_station =  - ((R * T) / (M * g)) * ln(P(h=0) / P(h=h_station))
            Where:
                P(h=0) is Pressure at Sea Level

    
    
    ++++++++++++++++++++++++++++++++++++++
    |||     Parameters of function     |||
    ++++++++++++++++++++++++++++++++++++++
    
        df: DataFrame of my study
        PreStat_col: Column of Pressure at Station in df
        PreSea_col: Column of Pressure at Sea Level in df
        Temp_col: Column of Temperature at Station in df 

    
    
    ++++++++++++++++++++++
    |||     Return     |||
    ++++++++++++++++++++++
    
        h_Station in meters
    
    '''
    
    # Physical Constent:-
    g = 9.80665
    M = 0.02896
    R = 8.314
    
    # Claculation:-
    h_station = (R * (df[Temp_col] + 273.15) / (M * g)) * np.log(df[PreSea_col] / df[PreStat_col])
    
    return h_station
    

In [186]:
df.columns

Index(['Date', 'AirTemperatureMeanDB', 'AirTemperatureMaxDB',
       'AirTemperatureMinDB', 'WindPrevailingDirection', 'WindMeanSpeed',
       'WindMaxDirection', 'WindMaxSpeed', 'RelativeHumidityMean',
       'SkyCovOktesMean', 'RelativeHumidityMax', 'RelativeHumidityMin',
       'PressureMinSeaLevel', 'PressureMinStationLevel',
       'PressureMaxStationLevel', 'PressureMaxSeaLevel',
       'PressureMeanStationLevel', 'PressureMeanSeaLevel', 'Year', 'Month',
       'Day', 'Hight'],
      dtype='object')

In [187]:
df['Hight'] = Hight(df, 'PressureMeanStationLevel', 'PressureMeanSeaLevel', 'AirTemperatureMeanDB')

C:\Users\Ibrahim Al Khurayyif\AppData\Local\Temp\ipykernel_20548\1045606646.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



##### The Mean of Hight and Standard error

In [188]:
# The Mean of Hight (meter):
Hight_mean = df['Hight'] .mean()
Hight_mean

237.89514715812638

In [189]:
# The Standard error of Hight (meter):
Hight_sem = df['Hight'] .sem()
Hight_sem

0.5223622113499298

#### Q2: Can you see if temperature and humidity have changed or not along 11 years ..?

### Plotting

In [190]:
fig = px.line(df, x='Date', y='AirTemperatureMeanDB')
fig.update_layout(
    title={
        'text': "Time Series of Air Tempreture mean",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.add_hline(y=31.761149874406946, annotation_text='Mean of Tempreture for Timeseries equals 31.761 °C ')
fig.update_annotations(font=dict(family="sans serif", size=25, color="black"))
fig.show()

In [191]:
fig = px.line(df, x='Date', y='RelativeHumidityMean')
fig.update_layout(
    title={
        'text': "Time Series of Relative Humidity mean",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.add_hline(y=45.98353335193971, line_color='red', 
              annotation_text='Mean of Relative Humidity of Timeseries equals 45.983 %',
             )

fig.update_annotations(font=dict(family="sans serif", size=25, color="black"))
fig.show()

In [222]:
fig = px.line(Temp_delta, x='Year', y=Temp_delta.drop(columns='Year').columns)
fig.update_layout(
    title={
        'text': "The Differnce between Tempreture mean of Month and Annual Tempreture mean",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.show()

In [219]:
fig = px.line(Humi_delta, x='Year', y=Humi_delta.drop(columns='Year').columns)
fig.update_layout(
    title={
        'text': "The Differnce between Humidity mean of Month and Annual Humidity mean",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.show()

From the graphs, there is fluctuation in the mean of temperature and mean of humidity levels, with no clear trend. The reason behind this is the time series of them is statonary process.

#### Q3. Is There correlation between Air Temperature and Humidity by Month ..?

Now I wil make a new dataframe for find carrolation between Tempreture and Humidity.

In [210]:
Temperature

Year        Jan        Feb        Mar        Apr        May        Jun  \
0  2008  23.560000  25.555172  29.323333  32.443333  34.273333  36.203333   
1  2009  25.813333  26.782143  27.296667  31.923333  33.880000  36.983333   
2  2010  25.326667  25.846429  28.540000  31.993333  34.363333  36.783333   
3  2011  24.010000  25.846429  26.880000  30.646667  34.356667  35.936667   
4  2012  23.553333  26.572414  27.296667  31.283333  35.510000  37.386667   
5  2013  25.276667  27.314286  29.583333  30.826667  34.986667  36.983333   
6  2014  25.650000  26.471429  28.780000  32.706667  34.053333  36.946667   
7  2015  24.416667  26.560714  29.403333  30.836667  35.196667  35.966667   
8  2016  24.396667  26.765517  31.540000  31.676667  35.530000  38.876667   
9  2017  27.516667  26.178571  29.386667  33.320000  35.313333  37.013333   

         Jul        Aug        Sep        Oct        Nov        Dec  
0  36.783333  36.166667  35.600000  32.430000  28.816667  26.706667  
1  36.316667  35.736667  34.920000  32.153333  28.380000  25.493333  
2  36.180000  36.666667  35.263333  34.290000  29.996667  26.833333  
3  36.853333  35.843333  34.703333  32.826667  27.500000  26.413333  
4  36.866667  36.163333  35.323333  32.576667  29.310000  26.460000  
5  36.070000  34.963333  35.466667  33.046667  29.636667  25.913333  
6  36.890000  36.523333  35.096667  32.596667  28.440000  27.310000  
7  36.986667  37.386667  36.150000  33.793333  29.580000  25.246667  
8  37.970000  36.753333  35.846667  33.236667  29.810000  28.240000  
9  37.476667  36.376667  36.070000  33.333333  29.420000  27.456667

In [211]:
df_comp = Temperature.merge(Humidity, on='Year')  # That is for merge between Tempreture and Humidity.
df_comp.columns = ['Year', 'Jan_Temp', 'Feb_Temp', 'Mar_Temp', 'Apr_Temp', 'May_Temp', 'Jun_Temp', 'Jul_Temp',
       'Aug_Temp', 'Sep_Temp', 'Oct_Temp', 'Nov_Temp', 'Dec_Temp', 'Jan_Humi', 'Feb_Humi', 'Mar_Humi',
       'Apr_Humi', 'May_Humi', 'Jun_Humi', 'Jul_Humi', 'Aug_Humi', 'Sep_Humi', 'Oct_Humi', 'Nov_Humi',
       'Dec_Humi']

In [212]:
df_comp = df_comp[[
    'Year', 'Jan_Temp',  'Jan_Humi', 'Feb_Temp', 'Feb_Humi', 'Mar_Temp', 'Mar_Humi', 'Apr_Temp', 'Apr_Humi',
    'May_Temp', 'May_Humi', 'Jun_Temp', 'Jun_Humi', 'Jul_Temp', 'Jul_Humi', 'Aug_Temp', 'Aug_Humi', 'Sep_Temp',
    'Sep_Humi', 'Oct_Temp', 'Oct_Humi', 'Nov_Temp', 'Nov_Humi', 'Dec_Temp',  'Dec_Humi'
]]

In [213]:
df_comp

Year   Jan_Temp   Jan_Humi   Feb_Temp   Feb_Humi   Mar_Temp   Mar_Humi  \
0  2008  23.560000  58.133333  25.555172  50.758621  29.323333  42.666667   
1  2009  25.813333  53.766667  26.782143  55.964286  27.296667  50.100000   
2  2010  25.326667  57.933333  25.846429  62.214286  28.540000  51.433333   
3  2011  24.010000  60.066667  25.846429  53.714286  26.880000  46.833333   
4  2012  23.553333  55.133333  26.572414  52.862069  27.296667  44.566667   
5  2013  25.276667  56.066667  27.314286  51.642857  29.583333  47.933333   
6  2014  25.650000  59.233333  26.471429  55.428571  28.780000  47.533333   
7  2015  24.416667  59.666667  26.560714  58.750000  29.403333  52.100000   
8  2016  24.396667  63.400000  26.765517  60.931034  31.540000  52.800000   
9  2017  27.516667  56.000000  26.178571  49.142857  29.386667  48.300000   

    Apr_Temp   Apr_Humi   May_Temp  ...   Aug_Temp   Aug_Humi   Sep_Temp  \
0  32.443333  39.766667  34.273333  ...  36.166667  31.766667  35.600000   
1  31.923333  38.066667  33.880000  ...  35.736667  41.000000  34.920000   
2  31.993333  46.400000  34.363333  ...  36.666667  43.266667  35.263333   
3  30.646667  39.833333  34.356667  ...  35.843333  36.700000  34.703333   
4  31.283333  40.233333  35.510000  ...  36.163333  33.833333  35.323333   
5  30.826667  39.800000  34.986667  ...  34.963333  43.033333  35.466667   
6  32.706667  42.266667  34.053333  ...  36.523333  39.200000  35.096667   
7  30.836667  41.733333  35.196667  ...  37.386667  44.000000  36.150000   
8  31.676667  52.600000  35.530000  ...  36.753333  48.100000  35.846667   
9  33.320000  41.133333  35.313333  ...  36.376667  45.233333  36.070000   

    Sep_Humi   Oct_Temp   Oct_Humi   Nov_Temp   Nov_Humi   Dec_Temp   Dec_Humi  
0  45.566667  32.430000  44.266667  28.816667  57.900000  26.706667  50.666667  
1  42.600000  32.153333  50.166667  28.380000  61.733333  25.493333  66.200000  
2  44.633333  34.290000  38.566667  29.996667  48.800000  26.833333  53.633333  
3  36.400000  32.826667  49.800000  27.500000  49.300000  26.413333  52.933333  
4  33.933333  32.576667  49.133333  29.310000  50.200000  26.460000  54.366667  
5  44.633333  33.046667  50.333333  29.636667  56.466667  25.913333  58.500000  
6  44.533333  32.596667  53.666667  28.440000  59.566667  27.310000  66.000000  
7  48.033333  33.793333  50.666667  29.580000  58.200000  25.246667  58.100000  
8  51.966667  33.236667  56.166667  29.810000  60.200000  28.240000  56.100000  
9  43.766667  33.333333  47.966667  29.420000  59.300000  27.456667  52.700000  

[10 rows x 25 columns]

In [214]:
corr_dict = {}
for i in range(12):
    corr_dict[df_comp.columns[2*i+1][0:3]] = df_comp[[df_comp.columns[0], 
                                                  df_comp.columns[2*i + 1],
                                                  df_comp.columns[2*i + 2]]].corr().values[1][2]

In [215]:
df_corr = pd.DataFrame.from_dict(corr_dict, orient='index').reset_index()
df_corr.columns = ['Month', 'corr(Temp, Humi)']

In [216]:
df_corr # Now I have DataFrame for Correlation. 

Month  corr(Temp, Humi)
0    Jan         -0.341264
1    Feb          0.064217
2    Mar          0.398964
3    Apr          0.070708
4    May          0.268934
5    Jun          0.026768
6    Jul         -0.143473
7    Aug          0.276425
8    Sep          0.599955
9    Oct         -0.387993
10   Nov          0.040335
11   Dec         -0.264178

In [217]:
fig = px.line(df_corr, x='Month', y='corr(Temp, Humi)')
fig.add_hrect(
        y0=0, y1=0.2, line_width=0, 
        fillcolor="green", opacity=0.1)
fig.add_hrect(
        y0=0.2, y1=0.5, line_width=0, 
        fillcolor="green", opacity=0.3)
fig.add_hrect(
        y0=0.5, y1=0.8, line_width=0, 
        fillcolor="green", opacity=0.5)
fig.add_hrect(
        y0=0.8, y1=1, line_width=0, 
        fillcolor="green", opacity=0.7)
fig.add_hrect(
        y0=0, y1=-0.2, line_width=0, 
        fillcolor="red", opacity=0.1)
fig.add_hrect(
        y0=-0.2, y1=-0.5, line_width=0, 
        fillcolor="red", opacity=0.3)
fig.add_hrect(
        y0=-0.5, y1=-0.8, line_width=0, 
        fillcolor="red", opacity=0.5)
fig.add_hrect(
        y0=-0.8, y1=-1, line_width=0, 
        fillcolor="red", opacity=0.7)
fig.update_layout(
    title={
        'text': "The Correlation between Temperature and Humidity for Month (2008 - 2017)",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.show()

From last fig, The correlation between  Temperature and Humidity is oscilated and is not strong, only September gives best correlation (It is 0.6).

In [218]:
Hight_mean, Hight_sem

(237.89514715812638, 0.5223622113499298)

# Results

1) I found the hight of station there is 237.89 +/- 0.52 above sea level.

2) there is fluctuation in the mean of temperature and mean of humidity levels, with no clear trend. The reason behind this is the time series of them is statonary process.

3) The correlation between Temperature and Humidity is oscilated and is not strong, only September gives best correlation (It is 0.6).